{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59152b9e-668c-44c6-b384-f059ebddaa90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initially, I downloaded the latest GENCODE gene annotation file, Release 45. My first step was to convert this into a CSV format, \n",
    "#which required adding headers to the columns based on the GENCODE GTF file's format description. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ab3f4e2-d4b5-48f8-ae84-fed1abd4f6aa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('firstfile_chromosomes_only1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f26d5e3-9e7d-490a-944e-4b173fd9445d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425351</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387460.2;Parent=ENSG00000210195.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425352</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000387460.2:1;Parent=ENST0000038...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425353</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000210196.2;gene_id=ENSG00000210196.2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425354</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387461.2;Parent=ENSG00000210196.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3425355</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000387461.2:1;Parent=ENST0000038...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3425356 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid   source        type  start    end score strand phase  \\\n",
       "0        chr1   HAVANA        gene  11869  14409     .      +     .   \n",
       "1        chr1   HAVANA  transcript  11869  14409     .      +     .   \n",
       "2        chr1   HAVANA        exon  11869  12227     .      +     .   \n",
       "3        chr1   HAVANA        exon  12613  12721     .      +     .   \n",
       "4        chr1   HAVANA        exon  13221  14409     .      +     .   \n",
       "...       ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "3425351  chrM  ENSEMBL  transcript  15888  15953     .      +     .   \n",
       "3425352  chrM  ENSEMBL        exon  15888  15953     .      +     .   \n",
       "3425353  chrM  ENSEMBL        gene  15956  16023     .      -     .   \n",
       "3425354  chrM  ENSEMBL  transcript  15956  16023     .      -     .   \n",
       "3425355  chrM  ENSEMBL        exon  15956  16023     .      -     .   \n",
       "\n",
       "                                                attributes  \n",
       "0        ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...  \n",
       "1        ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  \n",
       "2        ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3        ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4        ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  \n",
       "...                                                    ...  \n",
       "3425351  ID=ENST00000387460.2;Parent=ENSG00000210195.2;...  \n",
       "3425352  ID=exon:ENST00000387460.2:1;Parent=ENST0000038...  \n",
       "3425353  ID=ENSG00000210196.2;gene_id=ENSG00000210196.2...  \n",
       "3425354  ID=ENST00000387461.2;Parent=ENSG00000210196.2;...  \n",
       "3425355  ID=exon:ENST00000387461.2:1;Parent=ENST0000038...  \n",
       "\n",
       "[3425356 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9d8adb2-8d43-45ee-a5da-cf734242a9a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# I then filtered the dataset to select only the 'exons', using the 'type' column as my guide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d593da4d-7c3b-4463-bbd4-5d1541f20cd2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('chromosomes_only_exons_only_not_separated2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa3223ed-6e9f-4dd7-92bb-8b993afac72b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000450305.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650699</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000361681.2:1;Parent=ENST0000036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650700</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000387459.1:1;Parent=ENST0000038...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650701</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000361789.2:1;Parent=ENST0000036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650702</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000387460.2:1;Parent=ENST0000038...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650703</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000387461.2:1;Parent=ENST0000038...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1650704 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid   source  type  start    end score strand phase  \\\n",
       "0        chr1   HAVANA  exon  11869  12227     .      +     .   \n",
       "1        chr1   HAVANA  exon  12613  12721     .      +     .   \n",
       "2        chr1   HAVANA  exon  13221  14409     .      +     .   \n",
       "3        chr1   HAVANA  exon  12010  12057     .      +     .   \n",
       "4        chr1   HAVANA  exon  12179  12227     .      +     .   \n",
       "...       ...      ...   ...    ...    ...   ...    ...   ...   \n",
       "1650699  chrM  ENSEMBL  exon  14149  14673     .      -     .   \n",
       "1650700  chrM  ENSEMBL  exon  14674  14742     .      -     .   \n",
       "1650701  chrM  ENSEMBL  exon  14747  15887     .      +     .   \n",
       "1650702  chrM  ENSEMBL  exon  15888  15953     .      +     .   \n",
       "1650703  chrM  ENSEMBL  exon  15956  16023     .      -     .   \n",
       "\n",
       "                                                attributes  \n",
       "0        ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "1        ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "2        ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  \n",
       "3        ID=exon:ENST00000450305.2:1;Parent=ENST0000045...  \n",
       "4        ID=exon:ENST00000450305.2:2;Parent=ENST0000045...  \n",
       "...                                                    ...  \n",
       "1650699  ID=exon:ENST00000361681.2:1;Parent=ENST0000036...  \n",
       "1650700  ID=exon:ENST00000387459.1:1;Parent=ENST0000038...  \n",
       "1650701  ID=exon:ENST00000361789.2:1;Parent=ENST0000036...  \n",
       "1650702  ID=exon:ENST00000387460.2:1;Parent=ENST0000038...  \n",
       "1650703  ID=exon:ENST00000387461.2:1;Parent=ENST0000038...  \n",
       "\n",
       "[1650704 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6779ae64-6661-4a28-bc2b-1ae1b52bba52",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#The 'attributes' column contained essential, yet disorganized information, prompting me to sort it into distinct columns.\n",
    "# I accomplished this by exploiting the fundamental structure of Python dictionaries, segregating each piece of information based on its unique keys and values.\n",
    "import pandas as pd\n",
    "visual = pd.read_csv('chromosomes_only_exons_only_separated_attributes3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8653fc2b-68b0-426b-ba70-eba4d3dae622",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:3</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000450305.2:1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000450305.2:2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650699</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361681.2:1</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650700</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000387459.1:1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544476.1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650701</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361789.2:1</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650702</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000387460.2:1</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>...</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650703</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000387461.2:1</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>...</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1650704 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid   source  type  start    end score strand phase  \\\n",
       "0        chr1   HAVANA  exon  11869  12227     .      +     .   \n",
       "1        chr1   HAVANA  exon  12613  12721     .      +     .   \n",
       "2        chr1   HAVANA  exon  13221  14409     .      +     .   \n",
       "3        chr1   HAVANA  exon  12010  12057     .      +     .   \n",
       "4        chr1   HAVANA  exon  12179  12227     .      +     .   \n",
       "...       ...      ...   ...    ...    ...   ...    ...   ...   \n",
       "1650699  chrM  ENSEMBL  exon  14149  14673     .      -     .   \n",
       "1650700  chrM  ENSEMBL  exon  14674  14742     .      -     .   \n",
       "1650701  chrM  ENSEMBL  exon  14747  15887     .      +     .   \n",
       "1650702  chrM  ENSEMBL  exon  15888  15953     .      +     .   \n",
       "1650703  chrM  ENSEMBL  exon  15956  16023     .      -     .   \n",
       "\n",
       "                               ID             Parent  ...  \\\n",
       "0        exon:ENST00000456328.2:1  ENST00000456328.2  ...   \n",
       "1        exon:ENST00000456328.2:2  ENST00000456328.2  ...   \n",
       "2        exon:ENST00000456328.2:3  ENST00000456328.2  ...   \n",
       "3        exon:ENST00000450305.2:1  ENST00000450305.2  ...   \n",
       "4        exon:ENST00000450305.2:2  ENST00000450305.2  ...   \n",
       "...                           ...                ...  ...   \n",
       "1650699  exon:ENST00000361681.2:1  ENST00000361681.2  ...   \n",
       "1650700  exon:ENST00000387459.1:1  ENST00000387459.1  ...   \n",
       "1650701  exon:ENST00000361789.2:1  ENST00000361789.2  ...   \n",
       "1650702  exon:ENST00000387460.2:1  ENST00000387460.2  ...   \n",
       "1650703  exon:ENST00000387461.2:1  ENST00000387461.2  ...   \n",
       "\n",
       "                                  gene_type gene_name  \\\n",
       "0                                    lncRNA   DDX11L2   \n",
       "1                                    lncRNA   DDX11L2   \n",
       "2                                    lncRNA   DDX11L2   \n",
       "3        transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "4        transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "...                                     ...       ...   \n",
       "1650699                      protein_coding    MT-ND6   \n",
       "1650700                             Mt_tRNA     MT-TE   \n",
       "1650701                      protein_coding    MT-CYB   \n",
       "1650702                             Mt_tRNA     MT-TT   \n",
       "1650703                             Mt_tRNA     MT-TP   \n",
       "\n",
       "                            transcript_type transcript_name exon_number  \\\n",
       "0                                    lncRNA     DDX11L2-202           1   \n",
       "1                                    lncRNA     DDX11L2-202           2   \n",
       "2                                    lncRNA     DDX11L2-202           3   \n",
       "3        transcribed_unprocessed_pseudogene     DDX11L1-201           1   \n",
       "4        transcribed_unprocessed_pseudogene     DDX11L1-201           2   \n",
       "...                                     ...             ...         ...   \n",
       "1650699                      protein_coding      MT-ND6-201           1   \n",
       "1650700                             Mt_tRNA       MT-TE-201           1   \n",
       "1650701                      protein_coding      MT-CYB-201           1   \n",
       "1650702                             Mt_tRNA       MT-TT-201           1   \n",
       "1650703                             Mt_tRNA       MT-TP-201           1   \n",
       "\n",
       "                   exon_id  level transcript_support_level  \\\n",
       "0        ENSE00002234944.1      2                        1   \n",
       "1        ENSE00003582793.1      2                        1   \n",
       "2        ENSE00002312635.1      2                        1   \n",
       "3        ENSE00001948541.1      2                      NaN   \n",
       "4        ENSE00001671638.2      2                      NaN   \n",
       "...                    ...    ...                      ...   \n",
       "1650699  ENSE00001434974.2      3        ENSP00000354665.2   \n",
       "1650700  ENSE00001544476.1      3                      NaN   \n",
       "1650701  ENSE00001436074.2      3        ENSP00000354554.2   \n",
       "1650702  ENSE00001544475.2      3                      NaN   \n",
       "1650703  ENSE00001544473.2      3                      NaN   \n",
       "\n",
       "                             tag        havana_transcript  \n",
       "0        basic,Ensembl_canonical     OTTHUMT00000362751.1  \n",
       "1        basic,Ensembl_canonical     OTTHUMT00000362751.1  \n",
       "2        basic,Ensembl_canonical     OTTHUMT00000362751.1  \n",
       "3                     HGNC:37102  PGO:0000005,PGO:0000019  \n",
       "4                     HGNC:37102  PGO:0000005,PGO:0000019  \n",
       "...                          ...                      ...  \n",
       "1650699                      NaN                HGNC:7462  \n",
       "1650700                HGNC:7479  basic,Ensembl_canonical  \n",
       "1650701                      NaN                HGNC:7427  \n",
       "1650702                HGNC:7499  basic,Ensembl_canonical  \n",
       "1650703                HGNC:7494  basic,Ensembl_canonical  \n",
       "\n",
       "[1650704 rows x 22 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e50a2b62-37ba-4d73-9302-a53bf3cea0ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#After reorganizing the 'attributes' data, I applied a second filter to isolate only those entries marked as 'protein coding' in the 'gene_type' column, \n",
    "#Ensuring the dataset focused on the relevant gene types for my study.\n",
    "import pandas as pd\n",
    "visual = pd.read_csv('chromosomes_only_filteredby_proteincoding4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "83a3df11-c3d0-4fb7-a534-6842e357618c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:1</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:2</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:3</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000426406.4:1</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003989331.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000409316.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:31275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000332831.5:1</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004001351.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000329982.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:15079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388644</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361335.1:1</td>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001596097.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354728.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388645</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361381.2:1</td>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001666004.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354961.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388646</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361567.2:1</td>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001435330.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354813.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388647</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361681.2:1</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388648</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361789.2:1</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1388649 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid   source  type   start     end score strand phase  \\\n",
       "0        chr1   HAVANA  exon   65419   65433     .      +     .   \n",
       "1        chr1   HAVANA  exon   65520   65573     .      +     .   \n",
       "2        chr1   HAVANA  exon   69037   71585     .      +     .   \n",
       "3        chr1   HAVANA  exon  450740  451678     .      -     .   \n",
       "4        chr1   HAVANA  exon  685716  686654     .      -     .   \n",
       "...       ...      ...   ...     ...     ...   ...    ...   ...   \n",
       "1388644  chrM  ENSEMBL  exon   10470   10766     .      +     .   \n",
       "1388645  chrM  ENSEMBL  exon   10760   12137     .      +     .   \n",
       "1388646  chrM  ENSEMBL  exon   12337   14148     .      +     .   \n",
       "1388647  chrM  ENSEMBL  exon   14149   14673     .      -     .   \n",
       "1388648  chrM  ENSEMBL  exon   14747   15887     .      +     .   \n",
       "\n",
       "                               ID             Parent  ...       gene_type  \\\n",
       "0        exon:ENST00000641515.2:1  ENST00000641515.2  ...  protein_coding   \n",
       "1        exon:ENST00000641515.2:2  ENST00000641515.2  ...  protein_coding   \n",
       "2        exon:ENST00000641515.2:3  ENST00000641515.2  ...  protein_coding   \n",
       "3        exon:ENST00000426406.4:1  ENST00000426406.4  ...  protein_coding   \n",
       "4        exon:ENST00000332831.5:1  ENST00000332831.5  ...  protein_coding   \n",
       "...                           ...                ...  ...             ...   \n",
       "1388644  exon:ENST00000361335.1:1  ENST00000361335.1  ...  protein_coding   \n",
       "1388645  exon:ENST00000361381.2:1  ENST00000361381.2  ...  protein_coding   \n",
       "1388646  exon:ENST00000361567.2:1  ENST00000361567.2  ...  protein_coding   \n",
       "1388647  exon:ENST00000361681.2:1  ENST00000361681.2  ...  protein_coding   \n",
       "1388648  exon:ENST00000361789.2:1  ENST00000361789.2  ...  protein_coding   \n",
       "\n",
       "        gene_name transcript_type transcript_name exon_number  \\\n",
       "0           OR4F5  protein_coding       OR4F5-201           1   \n",
       "1           OR4F5  protein_coding       OR4F5-201           2   \n",
       "2           OR4F5  protein_coding       OR4F5-201           3   \n",
       "3          OR4F29  protein_coding      OR4F29-201           1   \n",
       "4          OR4F16  protein_coding      OR4F16-201           1   \n",
       "...           ...             ...             ...         ...   \n",
       "1388644   MT-ND4L  protein_coding     MT-ND4L-201           1   \n",
       "1388645    MT-ND4  protein_coding      MT-ND4-201           1   \n",
       "1388646    MT-ND5  protein_coding      MT-ND5-201           1   \n",
       "1388647    MT-ND6  protein_coding      MT-ND6-201           1   \n",
       "1388648    MT-CYB  protein_coding      MT-CYB-201           1   \n",
       "\n",
       "                   exon_id  level transcript_support_level         tag  \\\n",
       "0        ENSE00003812156.1      2        ENSP00000493376.2  HGNC:14825   \n",
       "1        ENSE00003813641.1      2        ENSP00000493376.2  HGNC:14825   \n",
       "2        ENSE00003813949.1      2        ENSP00000493376.2  HGNC:14825   \n",
       "3        ENSE00003989331.1      2        ENSP00000409316.1         NaN   \n",
       "4        ENSE00004001351.2      2        ENSP00000329982.2         NaN   \n",
       "...                    ...    ...                      ...         ...   \n",
       "1388644  ENSE00001596097.1      3        ENSP00000354728.1         NaN   \n",
       "1388645  ENSE00001666004.2      3        ENSP00000354961.2         NaN   \n",
       "1388646  ENSE00001435330.2      3        ENSP00000354813.2         NaN   \n",
       "1388647  ENSE00001434974.2      3        ENSP00000354665.2         NaN   \n",
       "1388648  ENSE00001436074.2      3        ENSP00000354554.2         NaN   \n",
       "\n",
       "                                         havana_transcript  \n",
       "0        RNA_Seq_supported_partial,basic,Ensembl_canoni...  \n",
       "1        RNA_Seq_supported_partial,basic,Ensembl_canoni...  \n",
       "2        RNA_Seq_supported_partial,basic,Ensembl_canoni...  \n",
       "3                                               HGNC:31275  \n",
       "4                                               HGNC:15079  \n",
       "...                                                    ...  \n",
       "1388644                                          HGNC:7460  \n",
       "1388645                                          HGNC:7459  \n",
       "1388646                                          HGNC:7461  \n",
       "1388647                                          HGNC:7462  \n",
       "1388648                                          HGNC:7427  \n",
       "\n",
       "[1388649 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a3b408c3-6627-4953-9927-af26389cc717",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed exon data saved to new_processed_exons_protein_coding.csv\n"
     ]
    }
   ],
   "source": [
    "#All in one step\n",
    "import pandas as pd\n",
    "\n",
    "def process_exons_split_attributes_and_filter(input_gff3_file, output_csv_file):\n",
    "    \"\"\"\n",
    "    Load a GFF3 file, filter for exons, split the 'attributes' column into separate columns,\n",
    "    filter for 'protein_coding' gene_type, and save the result to a CSV file.\n",
    "\n",
    "    Parameters:\n",
    "    input_gff3_file (str): The file path for the input GFF3 file, including file name.\n",
    "    output_csv_file (str): The file path for the output CSV file, including file name.\n",
    "    \"\"\"\n",
    "    # Loading the GFF3 file with specified headers\n",
    "    gff3_df = pd.read_csv(input_gff3_file, sep='\\t', comment='#', header=None,\n",
    "                          names=['seqid', 'source', 'type', 'start', 'end',\n",
    "                                 'score', 'strand', 'phase', 'attributes'])\n",
    "\n",
    "    # Filtering for exons\n",
    "    exon_df = gff3_df[gff3_df['type'] == 'exon']\n",
    "\n",
    "    # Spliting of the 'attributes' column into separate columns\n",
    "    attributes_df = exon_df['attributes'].str.split(';', expand=True)\n",
    "    attributes_expanded = pd.DataFrame()\n",
    "\n",
    "    for col in attributes_df.columns:\n",
    "        # Further split each attribute into key-value pairs\n",
    "        split_columns = attributes_df[col].str.split('=', expand=True)\n",
    "        if len(split_columns.columns) == 2:\n",
    "            # Use the first part as the key and the second part as the value\n",
    "            key = split_columns.iloc[:, 0].str.strip()\n",
    "            value = split_columns.iloc[:, 1]\n",
    "            attributes_expanded[key.iat[0]] = value\n",
    "\n",
    "    # To Merge the expanded attributes back into the exon DataFrame\n",
    "    data_final = pd.concat([exon_df.drop(columns=['attributes']), attributes_expanded], axis=1)\n",
    "\n",
    "    # To Filter for 'protein_coding' gene_type\n",
    "    protein_coding_df = data_final[data_final['gene_type'] == 'protein_coding']\n",
    "\n",
    "    # Saving the processed DataFrame to the specified CSV file\n",
    "    protein_coding_df.to_csv(output_csv_file, index=False)\n",
    "    print(f\"Processed exon data saved to {output_csv_file}\")\n",
    "\n",
    "\n",
    "input_gff3_file = 'gencode.v45.annotation.gff3'\n",
    "output_csv_file = 'new_processed_exons_protein_coding.csv'\n",
    "process_exons_split_attributes_and_filter(input_gff3_file, output_csv_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "035238b2-43de-4ba0-b05b-185d580cfc61",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('new_processed_exons_protein_coding.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "185aea78-99a4-471d-a09d-b27be932bf28",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:1</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:2</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:3</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000426406.4:1</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003989331.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000409316.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:31275</td>\n",
       "      <td>OTTHUMT00000007999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000332831.5:1</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004001351.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000329982.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>OTTHUMT00000007334.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388644</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361335.1:1</td>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>...</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001596097.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354728.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7460</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388645</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361381.2:1</td>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>...</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001666004.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354961.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7459</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388646</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361567.2:1</td>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>...</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001435330.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354813.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7461</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388647</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361681.2:1</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388648</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000361789.2:1</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>...</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1388649 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid   source  type   start     end score strand phase  \\\n",
       "0        chr1   HAVANA  exon   65419   65433     .      +     .   \n",
       "1        chr1   HAVANA  exon   65520   65573     .      +     .   \n",
       "2        chr1   HAVANA  exon   69037   71585     .      +     .   \n",
       "3        chr1   HAVANA  exon  450740  451678     .      -     .   \n",
       "4        chr1   HAVANA  exon  685716  686654     .      -     .   \n",
       "...       ...      ...   ...     ...     ...   ...    ...   ...   \n",
       "1388644  chrM  ENSEMBL  exon   10470   10766     .      +     .   \n",
       "1388645  chrM  ENSEMBL  exon   10760   12137     .      +     .   \n",
       "1388646  chrM  ENSEMBL  exon   12337   14148     .      +     .   \n",
       "1388647  chrM  ENSEMBL  exon   14149   14673     .      -     .   \n",
       "1388648  chrM  ENSEMBL  exon   14747   15887     .      +     .   \n",
       "\n",
       "                               ID             Parent  ... gene_name  \\\n",
       "0        exon:ENST00000641515.2:1  ENST00000641515.2  ...     OR4F5   \n",
       "1        exon:ENST00000641515.2:2  ENST00000641515.2  ...     OR4F5   \n",
       "2        exon:ENST00000641515.2:3  ENST00000641515.2  ...     OR4F5   \n",
       "3        exon:ENST00000426406.4:1  ENST00000426406.4  ...    OR4F29   \n",
       "4        exon:ENST00000332831.5:1  ENST00000332831.5  ...    OR4F16   \n",
       "...                           ...                ...  ...       ...   \n",
       "1388644  exon:ENST00000361335.1:1  ENST00000361335.1  ...   MT-ND4L   \n",
       "1388645  exon:ENST00000361381.2:1  ENST00000361381.2  ...    MT-ND4   \n",
       "1388646  exon:ENST00000361567.2:1  ENST00000361567.2  ...    MT-ND5   \n",
       "1388647  exon:ENST00000361681.2:1  ENST00000361681.2  ...    MT-ND6   \n",
       "1388648  exon:ENST00000361789.2:1  ENST00000361789.2  ...    MT-CYB   \n",
       "\n",
       "        transcript_type transcript_name exon_number            exon_id level  \\\n",
       "0        protein_coding       OR4F5-201           1  ENSE00003812156.1     2   \n",
       "1        protein_coding       OR4F5-201           2  ENSE00003813641.1     2   \n",
       "2        protein_coding       OR4F5-201           3  ENSE00003813949.1     2   \n",
       "3        protein_coding      OR4F29-201           1  ENSE00003989331.1     2   \n",
       "4        protein_coding      OR4F16-201           1  ENSE00004001351.2     2   \n",
       "...                 ...             ...         ...                ...   ...   \n",
       "1388644  protein_coding     MT-ND4L-201           1  ENSE00001596097.1     3   \n",
       "1388645  protein_coding      MT-ND4-201           1  ENSE00001666004.2     3   \n",
       "1388646  protein_coding      MT-ND5-201           1  ENSE00001435330.2     3   \n",
       "1388647  protein_coding      MT-ND6-201           1  ENSE00001434974.2     3   \n",
       "1388648  protein_coding      MT-CYB-201           1  ENSE00001436074.2     3   \n",
       "\n",
       "         transcript_support_level         tag  \\\n",
       "0               ENSP00000493376.2  HGNC:14825   \n",
       "1               ENSP00000493376.2  HGNC:14825   \n",
       "2               ENSP00000493376.2  HGNC:14825   \n",
       "3               ENSP00000409316.1         NaN   \n",
       "4               ENSP00000329982.2         NaN   \n",
       "...                           ...         ...   \n",
       "1388644         ENSP00000354728.1         NaN   \n",
       "1388645         ENSP00000354961.2         NaN   \n",
       "1388646         ENSP00000354813.2         NaN   \n",
       "1388647         ENSP00000354665.2         NaN   \n",
       "1388648         ENSP00000354554.2         NaN   \n",
       "\n",
       "                                         havana_transcript  \\\n",
       "0        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "1        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "2        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "3                                               HGNC:31275   \n",
       "4                                               HGNC:15079   \n",
       "...                                                    ...   \n",
       "1388644                                          HGNC:7460   \n",
       "1388645                                          HGNC:7459   \n",
       "1388646                                          HGNC:7461   \n",
       "1388647                                          HGNC:7462   \n",
       "1388648                                          HGNC:7427   \n",
       "\n",
       "                  Unnamed: 22  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3        OTTHUMT00000007999.3  \n",
       "4        OTTHUMT00000007334.3  \n",
       "...                       ...  \n",
       "1388644                   NaN  \n",
       "1388645                   NaN  \n",
       "1388646                   NaN  \n",
       "1388647                   NaN  \n",
       "1388648                   NaN  \n",
       "\n",
       "[1388649 rows x 23 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e760e88-abfb-4a4c-a9fa-ee5b2a701b26",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique transcript_IDs: 170476\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "csv_file_path = 'new_processed_exons_protein_coding.csv'  \n",
    "data = pd.read_csv(csv_file_path)\n",
    "\n",
    "# Count unique gene_IDs in the 'gene_id' column\n",
    "unique_gene_ids_count = data['Parent'].nunique()\n",
    "\n",
    "print(f\"Number of unique transcript_IDs: {unique_gene_ids_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "11920e0f-52a3-47cd-a782-1e5aac345cf0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique gene_IDs: 20049\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "csv_file_path = 'new_processed_exons_protein_coding.csv'  \n",
    "data = pd.read_csv(csv_file_path)\n",
    "\n",
    "# Count unique gene_IDs in the 'gene_id' column\n",
    "unique_gene_ids_count = data['gene_id'].nunique()\n",
    "\n",
    "print(f\"Number of unique gene_IDs: {unique_gene_ids_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f171d9fd-870f-4385-a7fc-8fadc7fb5643",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Sorting\n",
    "import pandas as pd\n",
    "\n",
    "exons_file = 'new_processed_exons_protein_coding.csv'\n",
    "exons_df = pd.read_csv(exons_file)\n",
    "exons_df.sort_values(['seqid','start'],inplace = True)\n",
    "exons_df.to_csv('new_last_gencode_exons_sorted.csv',index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8a46fe0f-c791-42f6-924d-1b1b78a6b7ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('new_last_gencode_exons_sorted.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b2d20fc9-f132-44b1-9a63-e096b5a7ad19",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:1</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:2</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:3</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000426406.4:1</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003989331.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000409316.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:31275</td>\n",
       "      <td>OTTHUMT00000007999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000332831.5:1</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004001351.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000329982.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>OTTHUMT00000007334.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388644</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711276.1:8</td>\n",
       "      <td>ENST00000711276.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-221</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00004015159.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388645</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711277.1:6</td>\n",
       "      <td>ENST00000711277.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-222</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00004015159.1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388646</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212230</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711278.1:8</td>\n",
       "      <td>ENST00000711278.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-223</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00004015162.1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>dotter_confirmed,non_canonical_TEC,not_best_in...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388647</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711284.1:3</td>\n",
       "      <td>ENST00000711284.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-229</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015172.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388648</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711285.1:3</td>\n",
       "      <td>ENST00000711285.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>WASH6P-230</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015172.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1388649 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid  source  type     start       end score strand phase  \\\n",
       "0        chr1  HAVANA  exon     65419     65433     .      +     .   \n",
       "1        chr1  HAVANA  exon     65520     65573     .      +     .   \n",
       "2        chr1  HAVANA  exon     69037     71585     .      +     .   \n",
       "3        chr1  HAVANA  exon    450740    451678     .      -     .   \n",
       "4        chr1  HAVANA  exon    685716    686654     .      -     .   \n",
       "...       ...     ...   ...       ...       ...   ...    ...   ...   \n",
       "1388644  chrY  HAVANA  exon  57211761  57212186     .      +     .   \n",
       "1388645  chrY  HAVANA  exon  57211761  57212186     .      +     .   \n",
       "1388646  chrY  HAVANA  exon  57211761  57212230     .      +     .   \n",
       "1388647  chrY  HAVANA  exon  57211761  57212074     .      +     .   \n",
       "1388648  chrY  HAVANA  exon  57211761  57212074     .      +     .   \n",
       "\n",
       "                               ID             Parent  ... gene_name  \\\n",
       "0        exon:ENST00000641515.2:1  ENST00000641515.2  ...     OR4F5   \n",
       "1        exon:ENST00000641515.2:2  ENST00000641515.2  ...     OR4F5   \n",
       "2        exon:ENST00000641515.2:3  ENST00000641515.2  ...     OR4F5   \n",
       "3        exon:ENST00000426406.4:1  ENST00000426406.4  ...    OR4F29   \n",
       "4        exon:ENST00000332831.5:1  ENST00000332831.5  ...    OR4F16   \n",
       "...                           ...                ...  ...       ...   \n",
       "1388644  exon:ENST00000711276.1:8  ENST00000711276.1  ...    WASH6P   \n",
       "1388645  exon:ENST00000711277.1:6  ENST00000711277.1  ...    WASH6P   \n",
       "1388646  exon:ENST00000711278.1:8  ENST00000711278.1  ...    WASH6P   \n",
       "1388647  exon:ENST00000711284.1:3  ENST00000711284.1  ...    WASH6P   \n",
       "1388648  exon:ENST00000711285.1:3  ENST00000711285.1  ...    WASH6P   \n",
       "\n",
       "                        transcript_type transcript_name exon_number  \\\n",
       "0                        protein_coding       OR4F5-201           1   \n",
       "1                        protein_coding       OR4F5-201           2   \n",
       "2                        protein_coding       OR4F5-201           3   \n",
       "3                        protein_coding      OR4F29-201           1   \n",
       "4                        protein_coding      OR4F16-201           1   \n",
       "...                                 ...             ...         ...   \n",
       "1388644                 retained_intron      WASH6P-221           8   \n",
       "1388645                 retained_intron      WASH6P-222           6   \n",
       "1388646                 retained_intron      WASH6P-223           8   \n",
       "1388647                 retained_intron      WASH6P-229           3   \n",
       "1388648  protein_coding_CDS_not_defined      WASH6P-230           3   \n",
       "\n",
       "                   exon_id level  transcript_support_level         tag  \\\n",
       "0        ENSE00003812156.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "1        ENSE00003813641.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "2        ENSE00003813949.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "3        ENSE00003989331.1     2         ENSP00000409316.1         NaN   \n",
       "4        ENSE00004001351.2     2         ENSP00000329982.2         NaN   \n",
       "...                    ...   ...                       ...         ...   \n",
       "1388644  ENSE00004015159.1     2                         1  HGNC:31685   \n",
       "1388645  ENSE00004015159.1     2                         5  HGNC:31685   \n",
       "1388646  ENSE00004015162.1     2                         5  HGNC:31685   \n",
       "1388647  ENSE00004015172.1     2                         1  HGNC:31685   \n",
       "1388648  ENSE00004015172.1     2                         2  HGNC:31685   \n",
       "\n",
       "                                         havana_transcript  \\\n",
       "0        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "1        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "2        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "3                                               HGNC:31275   \n",
       "4                                               HGNC:15079   \n",
       "...                                                    ...   \n",
       "1388644                               OTTHUMG00000022677.5   \n",
       "1388645                               OTTHUMG00000022677.5   \n",
       "1388646  dotter_confirmed,non_canonical_TEC,not_best_in...   \n",
       "1388647                               OTTHUMG00000022677.5   \n",
       "1388648                               OTTHUMG00000022677.5   \n",
       "\n",
       "                  Unnamed: 22  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3        OTTHUMT00000007999.3  \n",
       "4        OTTHUMT00000007334.3  \n",
       "...                       ...  \n",
       "1388644                   NaN  \n",
       "1388645                   NaN  \n",
       "1388646                   NaN  \n",
       "1388647                   NaN  \n",
       "1388648                   NaN  \n",
       "\n",
       "[1388649 rows x 23 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "db2efd61-b122-4d4b-a116-6ccfa52e3efb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed and sorted exon data saved to new_processed_exons_protein_coding111.csv\n"
     ]
    }
   ],
   "source": [
    "#All in the one step, (Headers, Filtering Exons & protein coding sequences, Sorting, and saving to a new csv file)\n",
    "import pandas as pd\n",
    "\n",
    "def process_exons_split_attributes_filter_and_sort(input_gff3_file, output_csv_file):\n",
    "    \"\"\"\n",
    "    Load a GFF3 file, filter for exons, split the 'attributes' column into separate columns,\n",
    "    filter for 'protein_coding' gene_type, sort the DataFrame, and save the result to a CSV file.\n",
    "\n",
    "    Parameters:\n",
    "    input_gff3_file (str): The file path for the input GFF3 file.\n",
    "    output_csv_file (str): The file path for the output CSV file.\n",
    "    \"\"\"\n",
    "    # Loading the GFF3 file with specified headers\n",
    "    gff3_df = pd.read_csv(input_gff3_file, sep='\\t', comment='#', header=None,\n",
    "                          names=['seqid', 'source', 'type', 'start', 'end',\n",
    "                                 'score', 'strand', 'phase', 'attributes'])\n",
    "\n",
    "    # Filtering for exons\n",
    "    exon_df = gff3_df[gff3_df['type'] == 'exon']\n",
    "\n",
    "    # Spliting the 'attributes' column into separate columns\n",
    "    attributes_df = exon_df['attributes'].str.split(';', expand=True)\n",
    "    attributes_expanded = pd.DataFrame()\n",
    "\n",
    "    for col in attributes_df.columns:\n",
    "        # Further split each attribute into key-value pairs\n",
    "        split_columns = attributes_df[col].str.split('=', expand=True)\n",
    "        if len(split_columns.columns) == 2:\n",
    "            # Uses the first part as the key and the second part as the value\n",
    "            key = split_columns.iloc[:, 0].str.strip()\n",
    "            value = split_columns.iloc[:, 1]\n",
    "            attributes_expanded[key.iat[0]] = value\n",
    "\n",
    "    # Merging the expanded attributes back into the exon DataFrame\n",
    "    data_final = pd.concat([exon_df.drop(columns=['attributes']), attributes_expanded], axis=1)\n",
    "\n",
    "    # Filtering for 'protein_coding' gene_type\n",
    "    protein_coding_df = data_final[data_final['gene_type'] == 'protein_coding']\n",
    "\n",
    "    # Sorting the DataFrame by 'seqid' and 'start'\n",
    "    sorted_df = protein_coding_df.sort_values(['seqid', 'start'])\n",
    "\n",
    "    # Saving the processed and sorted DataFrame to the specified CSV file\n",
    "    sorted_df.to_csv(output_csv_file, index=False)\n",
    "    print(f\"Processed and sorted exon data saved to {output_csv_file}\")\n",
    "\n",
    "\n",
    "input_gff3_file = 'gencode.v45.annotation.gff3'\n",
    "output_csv_file = 'new_processed_exons_protein_coding111.csv'  \n",
    "process_exons_split_attributes_filter_and_sort(input_gff3_file, output_csv_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dbc2ccf6-35ff-4cca-a0c3-54c9ec888ddb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('new_processed_exons_protein_coding111.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d1635a23-01d3-4786-8121-1ac17be375f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:1</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:2</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000641515.2:3</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td>HGNC:14825</td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000426406.4:1</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003989331.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000409316.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:31275</td>\n",
       "      <td>OTTHUMT00000007999.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000332831.5:1</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004001351.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000329982.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:15079</td>\n",
       "      <td>OTTHUMT00000007334.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388644</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711276.1:8</td>\n",
       "      <td>ENST00000711276.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-221</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00004015159.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388645</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711277.1:6</td>\n",
       "      <td>ENST00000711277.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-222</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00004015159.1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388646</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212230</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711278.1:8</td>\n",
       "      <td>ENST00000711278.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-223</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00004015162.1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>dotter_confirmed,non_canonical_TEC,not_best_in...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388647</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711284.1:3</td>\n",
       "      <td>ENST00000711284.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-229</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015172.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388648</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000711285.1:3</td>\n",
       "      <td>ENST00000711285.1</td>\n",
       "      <td>...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>WASH6P-230</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015172.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1388649 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid  source  type     start       end score strand phase  \\\n",
       "0        chr1  HAVANA  exon     65419     65433     .      +     .   \n",
       "1        chr1  HAVANA  exon     65520     65573     .      +     .   \n",
       "2        chr1  HAVANA  exon     69037     71585     .      +     .   \n",
       "3        chr1  HAVANA  exon    450740    451678     .      -     .   \n",
       "4        chr1  HAVANA  exon    685716    686654     .      -     .   \n",
       "...       ...     ...   ...       ...       ...   ...    ...   ...   \n",
       "1388644  chrY  HAVANA  exon  57211761  57212186     .      +     .   \n",
       "1388645  chrY  HAVANA  exon  57211761  57212186     .      +     .   \n",
       "1388646  chrY  HAVANA  exon  57211761  57212230     .      +     .   \n",
       "1388647  chrY  HAVANA  exon  57211761  57212074     .      +     .   \n",
       "1388648  chrY  HAVANA  exon  57211761  57212074     .      +     .   \n",
       "\n",
       "                               ID             Parent  ... gene_name  \\\n",
       "0        exon:ENST00000641515.2:1  ENST00000641515.2  ...     OR4F5   \n",
       "1        exon:ENST00000641515.2:2  ENST00000641515.2  ...     OR4F5   \n",
       "2        exon:ENST00000641515.2:3  ENST00000641515.2  ...     OR4F5   \n",
       "3        exon:ENST00000426406.4:1  ENST00000426406.4  ...    OR4F29   \n",
       "4        exon:ENST00000332831.5:1  ENST00000332831.5  ...    OR4F16   \n",
       "...                           ...                ...  ...       ...   \n",
       "1388644  exon:ENST00000711276.1:8  ENST00000711276.1  ...    WASH6P   \n",
       "1388645  exon:ENST00000711277.1:6  ENST00000711277.1  ...    WASH6P   \n",
       "1388646  exon:ENST00000711278.1:8  ENST00000711278.1  ...    WASH6P   \n",
       "1388647  exon:ENST00000711284.1:3  ENST00000711284.1  ...    WASH6P   \n",
       "1388648  exon:ENST00000711285.1:3  ENST00000711285.1  ...    WASH6P   \n",
       "\n",
       "                        transcript_type transcript_name exon_number  \\\n",
       "0                        protein_coding       OR4F5-201           1   \n",
       "1                        protein_coding       OR4F5-201           2   \n",
       "2                        protein_coding       OR4F5-201           3   \n",
       "3                        protein_coding      OR4F29-201           1   \n",
       "4                        protein_coding      OR4F16-201           1   \n",
       "...                                 ...             ...         ...   \n",
       "1388644                 retained_intron      WASH6P-221           8   \n",
       "1388645                 retained_intron      WASH6P-222           6   \n",
       "1388646                 retained_intron      WASH6P-223           8   \n",
       "1388647                 retained_intron      WASH6P-229           3   \n",
       "1388648  protein_coding_CDS_not_defined      WASH6P-230           3   \n",
       "\n",
       "                   exon_id level  transcript_support_level         tag  \\\n",
       "0        ENSE00003812156.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "1        ENSE00003813641.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "2        ENSE00003813949.1     2         ENSP00000493376.2  HGNC:14825   \n",
       "3        ENSE00003989331.1     2         ENSP00000409316.1         NaN   \n",
       "4        ENSE00004001351.2     2         ENSP00000329982.2         NaN   \n",
       "...                    ...   ...                       ...         ...   \n",
       "1388644  ENSE00004015159.1     2                         1  HGNC:31685   \n",
       "1388645  ENSE00004015159.1     2                         5  HGNC:31685   \n",
       "1388646  ENSE00004015162.1     2                         5  HGNC:31685   \n",
       "1388647  ENSE00004015172.1     2                         1  HGNC:31685   \n",
       "1388648  ENSE00004015172.1     2                         2  HGNC:31685   \n",
       "\n",
       "                                         havana_transcript  \\\n",
       "0        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "1        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "2        RNA_Seq_supported_partial,basic,Ensembl_canoni...   \n",
       "3                                               HGNC:31275   \n",
       "4                                               HGNC:15079   \n",
       "...                                                    ...   \n",
       "1388644                               OTTHUMG00000022677.5   \n",
       "1388645                               OTTHUMG00000022677.5   \n",
       "1388646  dotter_confirmed,non_canonical_TEC,not_best_in...   \n",
       "1388647                               OTTHUMG00000022677.5   \n",
       "1388648                               OTTHUMG00000022677.5   \n",
       "\n",
       "                  Unnamed: 22  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3        OTTHUMT00000007999.3  \n",
       "4        OTTHUMT00000007334.3  \n",
       "...                       ...  \n",
       "1388644                   NaN  \n",
       "1388645                   NaN  \n",
       "1388646                   NaN  \n",
       "1388647                   NaN  \n",
       "1388648                   NaN  \n",
       "\n",
       "[1388649 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d1bc7b8-04fb-4564-9603-afbea96bb73a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/uni-research'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a6f11f5-3257-4282-8493-860cbff87b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Next, I downloaded bed NarrowPeak files for all 139 unique RNA-binding proteins (RBPs), totaling 145 files with six duplicates among them. \n",
    "#I assigned headers to each file based on the ENCODE broadPeak format, which includes columns specifying the Start and End of each peak.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb89395-42d6-4dbb-a7d2-338083011a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#I then conducted a thorough check for overlaps between these peak locations and the exon data previously filtered, \n",
    "#ensuring the accuracy and relevance of our genomic intersections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59d36341-6ae5-4efc-8ed6-3d9aaaee7974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total RBP sites situated inside exons: 227\n",
      "input-rbp/ENCFF955LYD_AATF.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 52\n",
      "input-rbp/ENCFF781RUK_EEF2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 691\n",
      "input-rbp/ENCFF883KCN_MTPAP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1805\n",
      "input-rbp/ENCFF886XUO_SRSF1_1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3\n",
      "input-rbp/ENCFF313WDF_RPS11.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 598\n",
      "input-rbp/ENCFF635NZZ_CPSF6.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 64\n",
      "input-rbp/ENCFF822NWY_TAF15.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1663\n",
      "input-rbp/ENCFF900IOH_PCBP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2703\n",
      "input-rbp/ENCFF862ZPV_EFTUD2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 235\n",
      "input-rbp/ENCFF634UAC_EIF3G.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 139\n",
      "input-rbp/ENCFF623WGE_SLBP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 56\n",
      "input-rbp/ENCFF252CWU_DDX47.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 29\n",
      "input-rbp/ENCFF889AWX_HNRNPUL1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4180\n",
      "input-rbp/ENCFF111LMZ_FXR2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 38\n",
      "input-rbp/ENCFF328KJW_SSB.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 14\n",
      "input-rbp/ENCFF809CKN_LARP7.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 8429\n",
      "input-rbp/ENCFF520QYX_DDX43.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 328\n",
      "input-rbp/ENCFF128AKC_DHX30.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 97\n",
      "input-rbp/ENCFF295HRZ_SERBP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 351\n",
      "input-rbp/ENCFF786UOW_QKI.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 160\n",
      "input-rbp/ENCFF406MSI_HLTF.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 93\n",
      "input-rbp/ENCFF027UTH_KHDRBS1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 91\n",
      "input-rbp/ENCFF559HGK_PPIL4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 250\n",
      "input-rbp/ENCFF246EPM_MATR3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 56\n",
      "input-rbp/ENCFF125IDN_METTL1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 20\n",
      "input-rbp/ENCFF233PRB_NSUN2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 385\n",
      "input-rbp/ENCFF594EDV_SAFB2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2720\n",
      "input-rbp/ENCFF290DFO_U2AF2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 69\n",
      "input-rbp/ENCFF244JER_WDR43.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 294\n",
      "input-rbp/ENCFF886MLH_NCBP2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 281\n",
      "input-rbp/ENCFF023YVD_DGCR8.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 993\n",
      "input-rbp/ENCFF607ZRF_EWSR1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3784\n",
      "input-rbp/ENCFF223KVR_SRSF1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 453\n",
      "input-rbp/ENCFF205UTG_AKAP8L.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 702\n",
      "input-rbp/ENCFF696QWE_SLTM.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1387\n",
      "input-rbp/ENCFF366UJP_FXR1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3469\n",
      "input-rbp/ENCFF524ZZB_IGF2BP2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1283\n",
      "input-rbp/ENCFF797BHU_DROSHA.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 225\n",
      "input-rbp/ENCFF390NMJ_GARS.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 41\n",
      "input-rbp/ENCFF283WLS_APEX1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 96\n",
      "input-rbp/ENCFF618WYU_TBRG4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 988\n",
      "input-rbp/ENCFF206RIM_RBFOX2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 17052\n",
      "input-rbp/ENCFF365LFU_UCHL5.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 173\n",
      "input-rbp/ENCFF423SOR_FASTKD2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 34\n",
      "input-rbp/ENCFF404HHM_WRN.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 109\n",
      "input-rbp/ENCFF058OOW_ZRANB2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 132\n",
      "input-rbp/ENCFF799ERK_DDX51.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 208\n",
      "input-rbp/ENCFF198ULC_DDX52.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 32\n",
      "input-rbp/ENCFF392AEV_HNRNPA1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4964\n",
      "input-rbp/ENCFF650LMV_IGF2BP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 6467\n",
      "input-rbp/ENCFF087MAJ_DDX3X.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1115\n",
      "input-rbp/ENCFF037TVC_TARDBP2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 360\n",
      "input-rbp/ENCFF958GMP_XRCC6.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 130\n",
      "input-rbp/ENCFF554NIV_NIPBL.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1713\n",
      "input-rbp/ENCFF850ZSC_ELAC2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 5430\n",
      "input-rbp/ENCFF452MJL_PABPC4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 203\n",
      "input-rbp/ENCFF605AFN_MORC2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 19\n",
      "input-rbp/ENCFF241DVU_HNRNPU.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 7022\n",
      "input-rbp/ENCFF262ILV_BUD13.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1848\n",
      "input-rbp/ENCFF897MSA_EIF4E.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 302\n",
      "input-rbp/ENCFF007VPS_DDX1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 88\n",
      "input-rbp/ENCFF312VDN_RYBP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4334\n",
      "input-rbp/ENCFF630KXK_FMR1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 5782\n",
      "input-rbp/ENCFF858UKE_PRPF8.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 30\n",
      "input-rbp/ENCFF783INO_UTP18.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1475\n",
      "input-rbp/ENCFF904KTV_DDX6.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 28\n",
      "input-rbp/ENCFF114EUH_SDAD1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1154\n",
      "input-rbp/ENCFF278ZCU_GTF2F1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 11\n",
      "input-rbp/ENCFF128BDG_AARS.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 780\n",
      "input-rbp/ENCFF445ENC_HNRNPM.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 258\n",
      "input-rbp/ENCFF768WOD_ADAT1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 596\n",
      "input-rbp/ENCFF917CBK_HNRNPL.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 384\n",
      "input-rbp/ENCFF780OCY_SRSF7.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 168\n",
      "input-rbp/ENCFF794RJQ_TROVE2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 395\n",
      "input-rbp/ENCFF603WDI_MBNL1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2989\n",
      "input-rbp/ENCFF480NUZ_FMR1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 9075\n",
      "input-rbp/ENCFF537KWB_DDX24.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 12\n",
      "input-rbp/ENCFF723JLS_RPS10.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1001\n",
      "input-rbp/ENCFF918XJQ_HNRNPK.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 108\n",
      "input-rbp/ENCFF389PHE_RNF187.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 15738\n",
      "input-rbp/ENCFF949QGM_AQR.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 860\n",
      "input-rbp/ENCFF861KMV_FUS.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2\n",
      "input-rbp/ENCFF051EEW_SBDS.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 55\n",
      "input-rbp/ENCFF382URH_ZC3H8.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 374\n",
      "input-rbp/ENCFF385MJU_ILF3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2922\n",
      "input-rbp/ENCFF644LPB_METAP2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 108\n",
      "input-rbp/ENCFF953WTP_SAFB.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 14\n",
      "input-rbp/ENCFF941GJK_ABCF1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1188\n",
      "input-rbp/ENCFF327YTD_NOLC1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 148\n",
      "input-rbp/ENCFF167CDB_HNRNPC.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 788\n",
      "input-rbp/ENCFF907HNN_PTBP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 151\n",
      "input-rbp/ENCFF588ITW_PHF6.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 11485\n",
      "input-rbp/ENCFF597NVR_UPF1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 7182\n",
      "input-rbp/ENCFF649GEE_SF3B4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1713\n",
      "input-rbp/ENCFF730QRI_NONO.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 363\n",
      "input-rbp/ENCFF858CSS_EXOSC5.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 8\n",
      "input-rbp/ENCFF048RLZ_GNL3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1501\n",
      "input-rbp/ENCFF726PFJ_TRA2A.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 340\n",
      "input-rbp/ENCFF947CIN_LARP4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 401\n",
      "input-rbp/ENCFF858EHN_LSM11.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 14319\n",
      "input-rbp/ENCFF285IGP_ZNF622.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2519\n",
      "input-rbp/ENCFF094MQV_PUM1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 6179\n",
      "input-rbp/ENCFF061XNA_LIN28B.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2377\n",
      "input-rbp/ENCFF105HMJ_ZNF800.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2\n",
      "input-rbp/ENCFF028GAJ_UTP3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 17\n",
      "input-rbp/ENCFF887ARJ_SF3B1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1804\n",
      "input-rbp/ENCFF050VVI_GEMIN5.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 16781\n",
      "input-rbp/ENCFF746XLE_YBX3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 186\n",
      "input-rbp/ENCFF260ZOK_RPS6.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3179\n",
      "input-rbp/ENCFF566LNK_ELAVL1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1022\n",
      "input-rbp/ENCFF462WAI_DDX55.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 119\n",
      "input-rbp/ENCFF665BGJ_CPEB4.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4177\n",
      "input-rbp/ENCFF327DLL_GRWD1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 28\n",
      "input-rbp/ENCFF154AYH_NPM1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 594\n",
      "input-rbp/ENCFF593RED_TARDBP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 484\n",
      "input-rbp/ENCFF695OLO_XRN2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 938\n",
      "input-rbp/ENCFF640IHY_U2AF1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3\n",
      "input-rbp/ENCFF247NXL_PUS1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3030\n",
      "input-rbp/ENCFF918KMT_TIA1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2027\n",
      "input-rbp/ENCFF878VVA_CSTF2T.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2322\n",
      "input-rbp/ENCFF530HTL_RPS3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 593\n",
      "input-rbp/ENCFF207BXK_EIF4G2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4071\n",
      "input-rbp/ENCFF315HYU_FXR2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 12\n",
      "input-rbp/ENCFF239GPP_SUPV3L1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3458\n",
      "input-rbp/ENCFF031FMO_KHSRP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4127\n",
      "input-rbp/ENCFF158AOV_FAM120A.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 858\n",
      "input-rbp/ENCFF943XOU_SMNDC1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4035\n",
      "input-rbp/ENCFF211TRO_SND1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 200\n",
      "input-rbp/ENCFF932UXC_YWHAG.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1725\n",
      "input-rbp/ENCFF318RSO_HNRNPK.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 449\n",
      "input-rbp/ENCFF924OVC_FTO.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 46\n",
      "input-rbp/ENCFF399GGH_DDX42.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2927\n",
      "input-rbp/ENCFF136CWQ_FXR1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 4781\n",
      "input-rbp/ENCFF880MWQ_PUM2.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 919\n",
      "input-rbp/ENCFF472CCL_APOBEC3C.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 870\n",
      "input-rbp/ENCFF674BUM_GPKOW.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 762\n",
      "input-rbp/ENCFF663WYH_EXOSC10.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 992\n",
      "input-rbp/ENCFF258HOX_ZC3H11A.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 8006\n",
      "input-rbp/ENCFF597MMG_RBM15.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1222\n",
      "input-rbp/ENCFF033EFV_AKAP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 0\n",
      "input-rbp/ENCFF560RWD_WDR3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 19\n",
      "input-rbp/ENCFF378OYA_DDX21.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 865\n",
      "input-rbp/ENCFF810NKV_AGGF1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 92\n",
      "input-rbp/ENCFF039FGP_RBM22.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 162\n",
      "input-rbp/ENCFF781BNS_SRSF9.bed is done!\n",
      "****************************************\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def find_first_overlapping_exon(rbp_start, rbp_end, exons_starts, exons_ends):\n",
    "    low, high = 0, len(exons_starts) - 1\n",
    "    while low <= high:\n",
    "        mid = (low + high) // 2\n",
    "        exon_start, exon_end = exons_starts[mid], exons_ends[mid]\n",
    "\n",
    "        if exon_end < rbp_start:\n",
    "            low = mid + 1\n",
    "        elif exon_start > rbp_end:\n",
    "            high = mid - 1\n",
    "        else:\n",
    "            while mid > 0 and exons_ends[mid - 1] >= rbp_start:\n",
    "                mid -= 1\n",
    "            return mid\n",
    "    return None\n",
    "\n",
    "def process_overlap_findings(in_dir, out_dir, exon_annotation_csv):\n",
    "    # To Ensure the output directory exists\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    \n",
    "    # Loading exon annotations\n",
    "    exons_df = pd.read_csv(exon_annotation_csv)\n",
    "    exons_df.rename(columns={'seqid': 'chromosome'}, inplace=True)\n",
    "    column_names = ['chromosome', 'chromStart', 'chromEnd', 'signalValue', 'ID', 'Parent', 'gene_id', 'transcript_id', 'gene_type',\n",
    "                    'gene_name', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id']\n",
    "\n",
    "    files = os.listdir(in_dir)\n",
    "    for file in files:\n",
    "        if not file.endswith('.bed'):\n",
    "            continue  # Skip non-BED files\n",
    "\n",
    "        infile = os.path.join(in_dir, file)\n",
    "        rbp_sites_df = pd.read_csv(infile, sep='\\t', header=None)\n",
    "        rbp_sites_df.rename(columns={0: 'chr', 1: 'chromStart', 2: 'chromEnd', 3: 'name', 4: 'score', 5: 'strand', 6: 'signalValue', 7: 'pValue', 8: 'qValue', 9: 'peak'}, inplace=True)\n",
    "        rbp_sites_df.sort_values(['chr', 'chromStart'], inplace=True)\n",
    "        outfile = os.path.join(out_dir, file.replace(\".bed\", \"_out.csv\"))\n",
    "        rbp_sites_df.rename(columns={'chr': 'chromosome'}, inplace=True)\n",
    "\n",
    "        overlapping_rbp_sites = []\n",
    "\n",
    "        for chromosome in rbp_sites_df['chromosome'].unique():\n",
    "            if chromosome not in exons_df['chromosome'].unique():\n",
    "                continue\n",
    "\n",
    "            rbp_sites_chr = rbp_sites_df[rbp_sites_df['chromosome'] == chromosome]\n",
    "            exons_chr = exons_df[exons_df['chromosome'] == chromosome]\n",
    "\n",
    "            rbp_starts = rbp_sites_chr['chromStart'].to_numpy()\n",
    "            rbp_ends = rbp_sites_chr['chromEnd'].to_numpy()\n",
    "            rbp_signals = rbp_sites_chr['signalValue'].to_numpy()\n",
    "            exons_starts = exons_chr['start'].to_numpy()\n",
    "            exons_ends = exons_chr['end'].to_numpy()\n",
    "\n",
    "            for i, (rbp_start, rbp_end, rbp_signal) in enumerate(zip(rbp_starts, rbp_ends, rbp_signals)):\n",
    "                first_overlap_index = find_first_overlapping_exon(rbp_start, rbp_end, exons_starts, exons_ends)\n",
    "                if first_overlap_index is not None:\n",
    "                    for j in range(first_overlap_index, len(exons_starts)):\n",
    "                        exon_start, exon_end = exons_starts[j], exons_ends[j]\n",
    "                        if rbp_start <= exon_end and rbp_end >= exon_start:\n",
    "                            overlapping_rbp_sites.append({\n",
    "                                'chromosome': chromosome,\n",
    "                                'chromStart': rbp_start,\n",
    "                                'chromEnd': rbp_end,\n",
    "                                'signalValue': rbp_signal,\n",
    "                                **exons_chr.iloc[j][['ID', 'Parent', 'gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id']].to_dict()\n",
    "                            })\n",
    "                            break\n",
    "        \n",
    "        \n",
    "        overlapping_rbp_sites_df = pd.DataFrame(overlapping_rbp_sites)\n",
    "        overlapping_rbp_sites_df.to_csv(outfile, index=False)\n",
    "\n",
    "        print(f\"\\nTotal RBP sites situated inside exons: {len(overlapping_rbp_sites)}\")\n",
    "        print(f\"{infile} is done!\")\n",
    "        print(\"****************************************\")\n",
    "\n",
    "# Example usage\n",
    "in_dir = 'input-rbp'\n",
    "out_dir = 'overlap_RBPs_April'\n",
    "exon_annotation_csv = 'new_last_gencode_exons_sorted.csv'\n",
    "process_overlap_findings(in_dir, out_dir, exon_annotation_csv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7d075450-b19d-4b7e-8651-ae53b127b599",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total RBP sites situated inside exons: 93\n",
      "Test code/ENCFF027UTH_KHDRBS1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 281\n",
      "Test code/ENCFF023YVD_DGCR8.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 302\n",
      "Test code/ENCFF007VPS_DDX1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 2\n",
      "Test code/ENCFF028GAJ_UTP3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 3458\n",
      "Test code/ENCFF031FMO_KHSRP.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 1222\n",
      "Test code/ENCFF033EFV_AKAP1.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 0\n",
      "Test code/ENCFF560RWD_WDR3.bed is done!\n",
      "****************************************\n",
      "\n",
      "Total RBP sites situated inside exons: 92\n",
      "Test code/ENCFF039FGP_RBM22.bed is done!\n",
      "****************************************\n"
     ]
    }
   ],
   "source": [
    "#Better code, more consistency with the empty or zero inputs\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def find_first_overlapping_exon(rbp_start, rbp_end, exons_starts, exons_ends):\n",
    "    low, high = 0, len(exons_starts) - 1\n",
    "    while low <= high:\n",
    "        mid = (low + high) // 2\n",
    "        exon_start, exon_end = exons_starts[mid], exons_ends[mid]\n",
    "\n",
    "        if exon_end < rbp_start:\n",
    "            low = mid + 1\n",
    "        elif exon_start > rbp_end:\n",
    "            high = mid - 1\n",
    "        else:\n",
    "            while mid > 0 and exons_ends[mid - 1] >= rbp_start:\n",
    "                mid -= 1\n",
    "            return mid\n",
    "    return None\n",
    "\n",
    "def process_overlap_findings(in_dir, out_dir, exon_annotation_csv):\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    exons_df = pd.read_csv(exon_annotation_csv)\n",
    "    exons_df.rename(columns={'seqid': 'chromosome'}, inplace=True)\n",
    "    column_names = ['chromosome', 'chromStart', 'chromEnd', 'signalValue', 'ID', 'Parent', 'gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id']\n",
    "    files = os.listdir(in_dir)\n",
    "    for file in files:\n",
    "        if not file.endswith('.bed'):\n",
    "            continue\n",
    "        infile = os.path.join(in_dir, file)\n",
    "        rbp_sites_df = pd.read_csv(infile, sep='\\t', header=None)\n",
    "        rbp_sites_df.rename(columns={0: 'chr', 1: 'chromStart', 2: 'chromEnd', 3: 'name', 4: 'score', 5: 'strand', 6: 'signalValue', 7: 'pValue', 8: 'qValue', 9: 'peak'}, inplace=True)\n",
    "        rbp_sites_df.sort_values(['chr', 'chromStart'], inplace=True)\n",
    "        outfile = os.path.join(out_dir, file.replace(\".bed\", \"_out.csv\"))\n",
    "        rbp_sites_df.rename(columns={'chr': 'chromosome'}, inplace=True)\n",
    "        overlapping_rbp_sites = []\n",
    "        for chromosome in rbp_sites_df['chromosome'].unique():\n",
    "            if chromosome not in exons_df['chromosome'].unique():\n",
    "                continue\n",
    "            rbp_sites_chr = rbp_sites_df[rbp_sites_df['chromosome'] == chromosome]\n",
    "            exons_chr = exons_df[exons_df['chromosome'] == chromosome]\n",
    "            rbp_starts = rbp_sites_chr['chromStart'].to_numpy()\n",
    "            rbp_ends = rbp_sites_chr['chromEnd'].to_numpy()\n",
    "            rbp_signals = rbp_sites_chr['signalValue'].to_numpy()\n",
    "            exons_starts = exons_chr['start'].to_numpy()\n",
    "            exons_ends = exons_chr['end'].to_numpy()\n",
    "            for i, (rbp_start, rbp_end, rbp_signal) in enumerate(zip(rbp_starts, rbp_ends, rbp_signals)):\n",
    "                first_overlap_index = find_first_overlapping_exon(rbp_start, rbp_end, exons_starts, exons_ends)\n",
    "                if first_overlap_index is not None:\n",
    "                    for j in range(first_overlap_index, len(exons_starts)):\n",
    "                        exon_start, exon_end = exons_starts[j], exons_ends[j]\n",
    "                        if rbp_start <= exon_end and rbp_end >= exon_start:\n",
    "                            overlapping_rbp_sites.append({\n",
    "                                'chromosome': chromosome,\n",
    "                                'chromStart': rbp_start,\n",
    "                                'chromEnd': rbp_end,\n",
    "                                'signalValue': rbp_signal,\n",
    "                                **exons_chr.iloc[j][['ID', 'Parent', 'gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id']].to_dict()\n",
    "                            })\n",
    "                            break\n",
    "        # After finishing the loop:\n",
    "        if overlapping_rbp_sites:\n",
    "            overlapping_rbp_sites_df = pd.DataFrame(overlapping_rbp_sites)\n",
    "        else:\n",
    "            # Creating an empty DataFrame with the specified columns if no overlaps were found\n",
    "            overlapping_rbp_sites_df = pd.DataFrame(columns=column_names)        \n",
    "        # Saving to CSV, ensuring headers are included even if the DataFrame is empty\n",
    "        overlapping_rbp_sites_df.to_csv(outfile, index=False)\n",
    "        print(f\"\\nTotal RBP sites situated inside exons: {len(overlapping_rbp_sites)}\")\n",
    "        print(f\"{infile} is done!\")\n",
    "        print(\"****************************************\")\n",
    "in_dir = 'inpur-rbp/'\n",
    "out_dir = 'overlapp-RBPs'\n",
    "exon_annotation_csv = 'new_last_gencode_exons_sorted.csv'\n",
    "process_overlap_findings(in_dir, out_dir, exon_annotation_csv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7648a73d-5599-4c29-972e-8fa373590c70",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlapping entries have been saved to 'overlapping_entries_TARDBP.csv'.\n"
     ]
    }
   ],
   "source": [
    "#Of the 145 RNA-binding protein (RBP) experiments, six involved repeated studies. We analyzed each RBP's results individually and \n",
    "# then compiled a final dataset from only those findings that were consistent across repeated experiments (overlapped), ensuring the reliability of our analysis\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def merge_replicate_overlaps(replicate1_path, replicate2_path, output_path):\n",
    "    \"\"\"\n",
    "    Finds overlap regions between two files of biological replicates and saves the overlaps to a new CSV file.\n",
    "\n",
    "    Parameters:\n",
    "    replicate1_path (str): Path to the first CSV file of replicate data.\n",
    "    replicate2_path (str): Path to the second CSV file of replicate data.\n",
    "    output_path (str): Path to save the output CSV file containing the overlapping entries.\n",
    "    \"\"\"\n",
    "    # Load the CSV files for the biological replicates\n",
    "    replicate1_df = pd.read_csv(replicate1_path)\n",
    "    replicate2_df = pd.read_csv(replicate2_path)\n",
    "\n",
    "    # Ensuring the data is sorted can sometimes improve merge efficiency\n",
    "    replicate1_df.sort_values(by=['chromosome', 'chromStart', 'chromEnd'], inplace=True)\n",
    "    replicate2_df.sort_values(by=['chromosome', 'chromStart', 'chromEnd'], inplace=True)\n",
    "\n",
    "    # To Merge on 'chromosome' to ensure comparisons are made within the same chromosome\n",
    "    merged_df = pd.merge(replicate1_df, replicate2_df, on='chromosome', suffixes=('_r1', '_r2'))\n",
    "\n",
    "    # Vectorized check for overlaps\n",
    "    overlaps = merged_df[\n",
    "        (merged_df['chromStart_r1'] <= merged_df['chromEnd_r2']) &\n",
    "        (merged_df['chromEnd_r1'] >= merged_df['chromStart_r2'])\n",
    "    ]\n",
    "\n",
    "    # Extracting relevant information or cleaning up the DataFrame as needed\n",
    "\n",
    "    # Saving the overlaps to a new CSV file\n",
    "    overlaps.to_csv(output_path, index=False)\n",
    "    print(f\"Overlapping entries have been saved to '{output_path}'.\")\n",
    "\n",
    "replicate1_path = 'ENCFF037TVC_TARDBP2_out.csv'\n",
    "replicate2_path = 'ENCFF593RED_TARDBP_out.csv'\n",
    "output_path = 'overlapping_entries_TARDBP.csv'\n",
    "merge_replicate_overlaps(replicate1_path, replicate2_path, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "18194926-68af-45ed-9508-a97c6148a7a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('overlapping_entries_TARDBP.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6d090e20-4d62-4950-adf2-fd1ead784162",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>chromStart_r1</th>\n",
       "      <th>chromEnd_r1</th>\n",
       "      <th>signalValue_r1</th>\n",
       "      <th>exonSource_r1</th>\n",
       "      <th>exonStart_r1</th>\n",
       "      <th>exonEnd_r1</th>\n",
       "      <th>exonAttributes_r1</th>\n",
       "      <th>chromStart_r2</th>\n",
       "      <th>chromEnd_r2</th>\n",
       "      <th>signalValue_r2</th>\n",
       "      <th>exonSource_r2</th>\n",
       "      <th>exonStart_r2</th>\n",
       "      <th>exonEnd_r2</th>\n",
       "      <th>exonAttributes_r2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2372143</td>\n",
       "      <td>2372202</td>\n",
       "      <td>4.457885</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>2371165</td>\n",
       "      <td>2372591</td>\n",
       "      <td>ID=exon:ENST00000469374.5:5;Parent=ENST0000046...</td>\n",
       "      <td>2372124</td>\n",
       "      <td>2372292</td>\n",
       "      <td>4.860934</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>2371165</td>\n",
       "      <td>2372591</td>\n",
       "      <td>ID=exon:ENST00000469374.5:5;Parent=ENST0000046...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5873661</td>\n",
       "      <td>5873809</td>\n",
       "      <td>3.448159</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>5873252</td>\n",
       "      <td>5873841</td>\n",
       "      <td>ID=exon:ENST00000468253.1:1;Parent=ENST0000046...</td>\n",
       "      <td>5873697</td>\n",
       "      <td>5873944</td>\n",
       "      <td>5.853474</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>5873252</td>\n",
       "      <td>5873841</td>\n",
       "      <td>ID=exon:ENST00000468253.1:1;Parent=ENST0000046...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5873811</td>\n",
       "      <td>5873905</td>\n",
       "      <td>4.179357</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>5873252</td>\n",
       "      <td>5873841</td>\n",
       "      <td>ID=exon:ENST00000468253.1:1;Parent=ENST0000046...</td>\n",
       "      <td>5873697</td>\n",
       "      <td>5873944</td>\n",
       "      <td>5.853474</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>5873252</td>\n",
       "      <td>5873841</td>\n",
       "      <td>ID=exon:ENST00000468253.1:1;Parent=ENST0000046...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11023418</td>\n",
       "      <td>11023529</td>\n",
       "      <td>4.802249</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11023323</td>\n",
       "      <td>11023667</td>\n",
       "      <td>ID=exon:ENST00000617757.1:1;Parent=ENST0000061...</td>\n",
       "      <td>11023436</td>\n",
       "      <td>11023503</td>\n",
       "      <td>3.442259</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11023323</td>\n",
       "      <td>11023667</td>\n",
       "      <td>ID=exon:ENST00000617757.1:1;Parent=ENST0000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11023531</td>\n",
       "      <td>11023557</td>\n",
       "      <td>4.081849</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11023323</td>\n",
       "      <td>11023667</td>\n",
       "      <td>ID=exon:ENST00000617757.1:1;Parent=ENST0000061...</td>\n",
       "      <td>11023533</td>\n",
       "      <td>11023601</td>\n",
       "      <td>5.124133</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11023323</td>\n",
       "      <td>11023667</td>\n",
       "      <td>ID=exon:ENST00000617757.1:1;Parent=ENST0000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827234</td>\n",
       "      <td>73827314</td>\n",
       "      <td>4.940839</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>73826115</td>\n",
       "      <td>73827984</td>\n",
       "      <td>ID=exon:ENST00000648829.1:3;Parent=ENST0000064...</td>\n",
       "      <td>73827244</td>\n",
       "      <td>73827313</td>\n",
       "      <td>3.111660</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>73826115</td>\n",
       "      <td>73827984</td>\n",
       "      <td>ID=exon:ENST00000648829.1:3;Parent=ENST0000064...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX</td>\n",
       "      <td>120438635</td>\n",
       "      <td>120438718</td>\n",
       "      <td>5.601045</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>120436494</td>\n",
       "      <td>120439293</td>\n",
       "      <td>ID=exon:ENST00000371335.4:9;Parent=ENST0000037...</td>\n",
       "      <td>120438645</td>\n",
       "      <td>120438738</td>\n",
       "      <td>4.363074</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>120436494</td>\n",
       "      <td>120439293</td>\n",
       "      <td>ID=exon:ENST00000371335.4:9;Parent=ENST0000037...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX</td>\n",
       "      <td>131824760</td>\n",
       "      <td>131824828</td>\n",
       "      <td>4.726356</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>131823764</td>\n",
       "      <td>131825365</td>\n",
       "      <td>ID=exon:ENST00000657242.1:2;Parent=ENST0000065...</td>\n",
       "      <td>131824767</td>\n",
       "      <td>131824835</td>\n",
       "      <td>5.284272</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>131823764</td>\n",
       "      <td>131825365</td>\n",
       "      <td>ID=exon:ENST00000657242.1:2;Parent=ENST0000065...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153672920</td>\n",
       "      <td>153672975</td>\n",
       "      <td>4.463499</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>153672947</td>\n",
       "      <td>153673078</td>\n",
       "      <td>ID=exon:ENST00000462280.1:2;Parent=ENST0000046...</td>\n",
       "      <td>153672887</td>\n",
       "      <td>153673009</td>\n",
       "      <td>4.010528</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>153672947</td>\n",
       "      <td>153673078</td>\n",
       "      <td>ID=exon:ENST00000462280.1:2;Parent=ENST0000046...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154351158</td>\n",
       "      <td>154351240</td>\n",
       "      <td>3.969069</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>154351140</td>\n",
       "      <td>154351203</td>\n",
       "      <td>ID=exon:ENST00000498491.5:1;Parent=ENST0000049...</td>\n",
       "      <td>154351174</td>\n",
       "      <td>154351240</td>\n",
       "      <td>3.705882</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>154351140</td>\n",
       "      <td>154351203</td>\n",
       "      <td>ID=exon:ENST00000498491.5:1;Parent=ENST0000049...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>649 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome  chromStart_r1  chromEnd_r1  signalValue_r1 exonSource_r1  \\\n",
       "0         chr1        2372143      2372202        4.457885        HAVANA   \n",
       "1         chr1        5873661      5873809        3.448159        HAVANA   \n",
       "2         chr1        5873811      5873905        4.179357        HAVANA   \n",
       "3         chr1       11023418     11023529        4.802249        HAVANA   \n",
       "4         chr1       11023531     11023557        4.081849        HAVANA   \n",
       "..         ...            ...          ...             ...           ...   \n",
       "644       chrX       73827234     73827314        4.940839        HAVANA   \n",
       "645       chrX      120438635    120438718        5.601045        HAVANA   \n",
       "646       chrX      131824760    131824828        4.726356        HAVANA   \n",
       "647       chrX      153672920    153672975        4.463499        HAVANA   \n",
       "648       chrX      154351158    154351240        3.969069        HAVANA   \n",
       "\n",
       "     exonStart_r1  exonEnd_r1  \\\n",
       "0         2371165     2372591   \n",
       "1         5873252     5873841   \n",
       "2         5873252     5873841   \n",
       "3        11023323    11023667   \n",
       "4        11023323    11023667   \n",
       "..            ...         ...   \n",
       "644      73826115    73827984   \n",
       "645     120436494   120439293   \n",
       "646     131823764   131825365   \n",
       "647     153672947   153673078   \n",
       "648     154351140   154351203   \n",
       "\n",
       "                                     exonAttributes_r1  chromStart_r2  \\\n",
       "0    ID=exon:ENST00000469374.5:5;Parent=ENST0000046...        2372124   \n",
       "1    ID=exon:ENST00000468253.1:1;Parent=ENST0000046...        5873697   \n",
       "2    ID=exon:ENST00000468253.1:1;Parent=ENST0000046...        5873697   \n",
       "3    ID=exon:ENST00000617757.1:1;Parent=ENST0000061...       11023436   \n",
       "4    ID=exon:ENST00000617757.1:1;Parent=ENST0000061...       11023533   \n",
       "..                                                 ...            ...   \n",
       "644  ID=exon:ENST00000648829.1:3;Parent=ENST0000064...       73827244   \n",
       "645  ID=exon:ENST00000371335.4:9;Parent=ENST0000037...      120438645   \n",
       "646  ID=exon:ENST00000657242.1:2;Parent=ENST0000065...      131824767   \n",
       "647  ID=exon:ENST00000462280.1:2;Parent=ENST0000046...      153672887   \n",
       "648  ID=exon:ENST00000498491.5:1;Parent=ENST0000049...      154351174   \n",
       "\n",
       "     chromEnd_r2  signalValue_r2 exonSource_r2  exonStart_r2  exonEnd_r2  \\\n",
       "0        2372292        4.860934        HAVANA       2371165     2372591   \n",
       "1        5873944        5.853474        HAVANA       5873252     5873841   \n",
       "2        5873944        5.853474        HAVANA       5873252     5873841   \n",
       "3       11023503        3.442259        HAVANA      11023323    11023667   \n",
       "4       11023601        5.124133        HAVANA      11023323    11023667   \n",
       "..           ...             ...           ...           ...         ...   \n",
       "644     73827313        3.111660        HAVANA      73826115    73827984   \n",
       "645    120438738        4.363074        HAVANA     120436494   120439293   \n",
       "646    131824835        5.284272        HAVANA     131823764   131825365   \n",
       "647    153673009        4.010528        HAVANA     153672947   153673078   \n",
       "648    154351240        3.705882        HAVANA     154351140   154351203   \n",
       "\n",
       "                                     exonAttributes_r2  \n",
       "0    ID=exon:ENST00000469374.5:5;Parent=ENST0000046...  \n",
       "1    ID=exon:ENST00000468253.1:1;Parent=ENST0000046...  \n",
       "2    ID=exon:ENST00000468253.1:1;Parent=ENST0000046...  \n",
       "3    ID=exon:ENST00000617757.1:1;Parent=ENST0000061...  \n",
       "4    ID=exon:ENST00000617757.1:1;Parent=ENST0000061...  \n",
       "..                                                 ...  \n",
       "644  ID=exon:ENST00000648829.1:3;Parent=ENST0000064...  \n",
       "645  ID=exon:ENST00000371335.4:9;Parent=ENST0000037...  \n",
       "646  ID=exon:ENST00000657242.1:2;Parent=ENST0000065...  \n",
       "647  ID=exon:ENST00000462280.1:2;Parent=ENST0000046...  \n",
       "648  ID=exon:ENST00000498491.5:1;Parent=ENST0000049...  \n",
       "\n",
       "[649 rows x 15 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7c1a996a-f8ab-4c5b-9ee8-688359a07df1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing beeeiiin/overlapping_entries_TARDBP.csv...\n",
      "Updated data saved to beeeiiin/overlapping_entries_TARDBP.csv\n",
      "All files have been processed.\n"
     ]
    }
   ],
   "source": [
    "# I used the mean of SignalValue of replicate1 and 2\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def process_files_and_update_signal_values(input_folder_path, output_folder_path):\n",
    "    \"\"\"\n",
    "    Processes CSV files in a specified folder by calculating the mean of signalValue_r1 and signalValue_r2,\n",
    "    keeps only the first 14 columns, renames columns by removing '_r1' suffix, and saves the processed files\n",
    "    to a new output folder.\n",
    "\n",
    "    Parameters:\n",
    "    - input_folder_path: Path to the folder containing the input CSV files.\n",
    "    - output_folder_path: Path to the folder where processed CSV files will be saved.\n",
    "    \"\"\"\n",
    "    # Ensuring the output folder exists\n",
    "    os.makedirs(output_folder_path, exist_ok=True)\n",
    "\n",
    "    # Iterate through each file in the input folder\n",
    "    for filename in os.listdir(input_folder_path):\n",
    "        if filename.endswith(\".csv\"):\n",
    "            input_file_path = os.path.join(input_folder_path, filename)\n",
    "            output_file_path = os.path.join(output_folder_path, filename)\n",
    "            print(f\"Processing {input_file_path}...\")\n",
    "\n",
    "            # Read the CSV file\n",
    "            df = pd.read_csv(input_file_path)\n",
    "\n",
    "            # Calculating the mean of signalValue_r1 and signalValue_r2, update signalValue_r1\n",
    "            df['signalValue_r1'] = df[['signalValue_r1', 'signalValue_r2']].mean(axis=1)\n",
    "\n",
    "            # Keeping only the first 14 columns\n",
    "            df = df.iloc[:, :14]\n",
    "\n",
    "            # Removing \"_r1\" from the column headers\n",
    "            df.columns = [col.replace(\"_r1\", \"\") for col in df.columns]\n",
    "\n",
    "            # Saving the processed DataFrame to a new file in the output folder\n",
    "            df.to_csv(output_file_path, index=False)\n",
    "            \n",
    "            print(f\"Updated data saved to {output_file_path}\")\n",
    "\n",
    "input_folder_path = 'beeeiiin/'  \n",
    "output_folder_path = 'beeeiiin/'  \n",
    "process_files_and_update_signal_values(input_folder_path, output_folder_path)\n",
    "\n",
    "print(\"All files have been processed.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "62fdd380-0588-4da6-9829-79cb49eeb854",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed and saved: Birthdayparty1/summed_ENCFF566LNK_ELAVL1_out.csv\n",
      "Processed and saved: Birthdayparty1/summed_ENCFF797BHU_DROSHA_out.csv\n",
      "Processed and saved: Birthdayparty1/summed_ENCFF786UOW_QKI_out.csv\n",
      "Found empty file: ENCFF560RWD_WDR3_out.csv. Header-only file created: Birthdayparty1/summed_ENCFF560RWD_WDR3_out.csv\n",
      "Processed and saved: Birthdayparty1/summed_ENCFF768WOD_ADAT1_out.csv\n",
      "Processed and saved: Birthdayparty1/summed_ENCFF941GJK_ABCF1_out.csv\n",
      "Processed and saved: Birthdayparty1/summed_ENCFF730QRI_NONO_out.csv\n"
     ]
    }
   ],
   "source": [
    "#This function, sum_signal_values_by_gene_id, correctly process each CSV file by summing up signalValue for each unique gene_id.\n",
    "#It creates an output file for each input file, where each output file contains two columns: gene_id and the corresponding sum of signalValue. \n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def sum_signal_values_by_gene_id(input_dir, output_dir):\n",
    "    \"\"\"\n",
    "    For each CSV file in the input directory, this function sums the 'signalValue' for each unique 'gene_id'.\n",
    "    It then writes these sums along with the 'gene_id' to a new CSV file in the output directory.\n",
    "    If an input file is empty, an output file with just the headers 'gene_id' and 'signalValue' is created.\n",
    "\n",
    "    Parameters:\n",
    "    input_dir (str): The directory containing the input CSV files.\n",
    "    output_dir (str): The directory where the output CSV files will be saved.\n",
    "    \"\"\"\n",
    "    # To Ensure the output directory exists\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    # List all CSV files in the input directory\n",
    "    rbp_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "    for rbp_file in rbp_files:\n",
    "        input_file_path = os.path.join(input_dir, rbp_file)\n",
    "        output_file_path = os.path.join(output_dir, f\"summed_{rbp_file}\")\n",
    "        \n",
    "        # Attempt to load the current RBP file\n",
    "        df = pd.read_csv(input_file_path)\n",
    "\n",
    "        # If the DataFrame is empty, create an output file with only headers\n",
    "        if df.empty:\n",
    "            empty_df = pd.DataFrame(columns=['gene_id', 'signalValue'])\n",
    "            empty_df.to_csv(output_file_path, index=False)\n",
    "            print(f\"Found empty file: {rbp_file}. Header-only file created: {output_file_path}\")\n",
    "            continue\n",
    "\n",
    "        # Sum 'signalValue's for each unique 'gene_id'\n",
    "        summed_df = df.groupby('gene_id')['signalValue'].sum().reset_index()\n",
    "\n",
    "        # Saving the summed data to a new CSV file\n",
    "        summed_df.to_csv(output_file_path, index=False)\n",
    "        print(f\"Processed and saved: {output_file_path}\")\n",
    "\n",
    "\n",
    "input_dir = 'Birthday/'  \n",
    "output_dir = 'Birthdayparty1'  \n",
    "sum_signal_values_by_gene_id(input_dir, output_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e8d85bd3-b997-44cd-aeb8-95f399019f92",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('summed_ENCFF730QRI_NONO_out.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b4d29b1b-fb32-4dd4-aa8d-c44a5acf312f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signalValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>7.782022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000007392.17</td>\n",
       "      <td>3.811146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000008083.14</td>\n",
       "      <td>9.624000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000008130.15</td>\n",
       "      <td>10.227331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000010310.9</td>\n",
       "      <td>9.093824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>ENSG00000284874.1</td>\n",
       "      <td>4.371654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>ENSG00000286192.1</td>\n",
       "      <td>12.115268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>ENSG00000288635.1</td>\n",
       "      <td>4.320107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>ENSG00000288649.2</td>\n",
       "      <td>58.563762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>ENSG00000291317.1</td>\n",
       "      <td>4.379898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>633 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id  signalValue\n",
       "0     ENSG00000004478.8     7.782022\n",
       "1    ENSG00000007392.17     3.811146\n",
       "2    ENSG00000008083.14     9.624000\n",
       "3    ENSG00000008130.15    10.227331\n",
       "4     ENSG00000010310.9     9.093824\n",
       "..                  ...          ...\n",
       "628   ENSG00000284874.1     4.371654\n",
       "629   ENSG00000286192.1    12.115268\n",
       "630   ENSG00000288635.1     4.320107\n",
       "631   ENSG00000288649.2    58.563762\n",
       "632   ENSG00000291317.1     4.379898\n",
       "\n",
       "[633 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e380f960-4ddb-4c69-b1db-991e23baa532",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To achieve the code that converts the CSV annotation file into the GFF3 format within a function, \n",
    "#we defined a function named csv_to_gff3. This function takes the path to the input CSV file and the desired output GFF3 file path as parameters.\n",
    "#This conversion is essential for creating a local GFF file filtered by exons of protein-coding genes,\n",
    "#allowing us to align Ribo-seq data precisely against these gene annotations. This ensures accurate measurement of translation activity, crucial for our analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0d176365-779f-4dd4-8da0-edda9b88ba7c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GFF3 file saved to test-proteinfiltered_output.gff3\n",
      "GFF3 file compressed and saved to test-proteinfiltered_output.gff3.gz\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "import shutil\n",
    "import os\n",
    "\n",
    "def csv_to_gff3(input_csv_path, output_gff3_path):\n",
    "    \"\"\"\n",
    "    Converts a CSV file containing gene annotations into a GFF3 format file.\n",
    "    \n",
    "    Parameters:\n",
    "    input_csv_path (str): The file path to the input CSV file with annotations.\n",
    "    output_gff3_path (str): The file path where the output GFF3 file should be saved.\n",
    "    \"\"\"\n",
    "    # Load the CSV file\n",
    "    data = pd.read_csv(input_csv_path)\n",
    "\n",
    "    # Construct the 'attributes' column with all necessary details\n",
    "    data['attributes'] = data.apply(\n",
    "        lambda x: f\"ID={x['ID']};Parent={x['Parent']};gene_id={x['gene_id']};\"\n",
    "                  f\"transcript_id={x['transcript_id']};gene_type={x['gene_type']};\"\n",
    "                  f\"gene_name={x['gene_name']};transcript_type={x['transcript_type']};\"\n",
    "                  f\"transcript_name={x['transcript_name']};exon_number={x['exon_number']};\"\n",
    "                  f\"exon_id={x['exon_id']};level={x['level']};\"\n",
    "                  f\"transcript_support_level={x.get('transcript_support_level', '')};\"\n",
    "                  f\"tag={x.get('tag', '')};havana_transcript={x.get('havana_transcript', '')}\",\n",
    "        axis=1\n",
    "    ).str.replace(';;', ';')  # Clean up potential trailing semicolons\n",
    "\n",
    "    # Format the DataFrame for GFF3\n",
    "    gff3_columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    gff3_data = data[gff3_columns].fillna({\n",
    "        'source': '.',  # Fill in with actual data or a placeholder\n",
    "        'score': '.',\n",
    "        'strand': '+',  # Adjust based on your data\n",
    "        'phase': '.'\n",
    "    })\n",
    "\n",
    "    # Convert 'start' and 'end' to integers\n",
    "    gff3_data['start'] = gff3_data['start'].astype(int)\n",
    "    gff3_data['end'] = gff3_data['end'].astype(int)\n",
    "\n",
    "    # Save to GFF3 file\n",
    "    gff3_data.to_csv(output_gff3_path, sep='\\t', index=False, header=False, quoting=3)\n",
    "\n",
    "    print(f\"GFF3 file saved to {output_gff3_path}\")\n",
    "    gz_file_path = output_gff3_path + '.gz'\n",
    "    with open(output_gff3_path, 'rb') as f_in:\n",
    "        with gzip.open(gz_file_path, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    print(f\"GFF3 file compressed and saved to {gz_file_path}\")\n",
    "    \n",
    "# Example usage\n",
    "input_csv_path = 'new_last_gencode_exons_sorted.csv'  # Update this to your actual file path\n",
    "output_gff3_path = 'test-proteinfiltered_output.gff3'  # Update this to your desired output path\n",
    "csv_to_gff3(input_csv_path, output_gff3_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "505073a8-85be-48af-aa17-b1744f143456",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/uni-research'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1931569-0aac-4d06-9019-67d5ddb761fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Then we generate the csv file from downloaded gene length file from https://usegalaxy.org/ \n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "input_file_path = 'Galaxy63-[featureCounts_on_data_55_and_data_2__Feature_lengths].tabular'\n",
    "\n",
    "\n",
    "output_file_path = 'Featurecounts_proteinfiltered_length_output_file.csv'\n",
    "\n",
    "# Loading the .tabular file using pandas\n",
    "# The separator for .tabular files is typically a tab ('\\t'), similar to TSV files\n",
    "data = pd.read_csv(input_file_path, sep='\\t')\n",
    "\n",
    "# Saveing the DataFrame to a CSV file\n",
    "# This will use a comma as the default delimiter\n",
    "data.to_csv(output_file_path, index=False)\n",
    "\n",
    "print(f\"Converted file saved to {output_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "34692821-434f-4385-9ed8-9770cfb2b3d7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('Featurecounts_proteinfiltered_length_output_file.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "514266f2-6f72-4eb0-8b8e-7d4aea41b18c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>2618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>4172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>5540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20044</th>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20045</th>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>1378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20046</th>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>1812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20047</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20048</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20049 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Geneid  Length\n",
       "0       ENSG00000186092.7    2618\n",
       "1       ENSG00000284733.2     939\n",
       "2       ENSG00000284662.2     939\n",
       "3      ENSG00000187634.13    4172\n",
       "4      ENSG00000188976.11    5540\n",
       "...                   ...     ...\n",
       "20044   ENSG00000212907.2     297\n",
       "20045   ENSG00000198886.2    1378\n",
       "20046   ENSG00000198786.2    1812\n",
       "20047   ENSG00000198695.2     525\n",
       "20048   ENSG00000198727.2    1141\n",
       "\n",
       "[20049 rows x 2 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18b749a4-d43f-465c-b954-67db5502edae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combining two tabular files from Ribo profiles (two replicates(replicate1 and replicate2):output of FeatureCounts)\n",
    "import pandas as pd\n",
    "\n",
    "# Assuming 'featureCounts.tabular' is your file name\n",
    "replicate1 = pd.read_csv('Galaxy61-[featureCounts_on_data_55_and_data_2__Counts].tabular', sep='\\t', comment='#', index_col=0)\n",
    "replicate2 = pd.read_csv('Galaxy64-[featureCounts_on_data_55_and_data_5__Counts].tabular', sep='\\t', comment='#', index_col=0)\n",
    "\n",
    "# Combining the counts from the last column of each file\n",
    "combined = pd.concat([replicate1.iloc[:, -1], replicate2.iloc[:, -1]], axis=1)\n",
    "combined.columns = ['Replicate1', 'Replicate2']\n",
    "\n",
    "# Save combined data\n",
    "combined.to_csv('combined_featureCounts_chr_proteinfiltered_shScramble.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aa17cce-712c-4b90-bc1f-32c278f0a72d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting the average from counts of two replicates\n",
    "import pandas as pd\n",
    "\n",
    "combined_featurecounts_path = 'combined_featureCounts_chr_proteinfiltered_shScramble.csv'\n",
    "combined_featurecounts = pd.read_csv(combined_featurecounts_path)\n",
    "\n",
    "combined_featurecounts['Mean_Counts'] = combined_featurecounts.iloc[:, 1:3].mean(axis=1).astype(int)\n",
    "\n",
    "combined_featurecounts.to_csv('mean_combined_featureCounts_chr_proteinfiltered_shScramble.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94ec12c2-1fd2-43cb-bd38-3a7f4d091c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting the average from counts of two replicates\n",
    "import pandas as pd\n",
    "\n",
    "combined_featurecounts_path = 'combined_featureCounts_chr_proteinfiltered_shScramble.csv'\n",
    "combined_featurecounts = pd.read_csv(combined_featurecounts_path)\n",
    "\n",
    "combined_featurecounts['Mean_Counts'] = combined_featurecounts.iloc[:, 1:3].mean(axis=1).astype(int)\n",
    "\n",
    "combined_featurecounts.to_csv('mean_combined_featureCounts_chr_proteinfiltered_shScramble.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ed756cf-e2ef-481f-847d-cd0f0d853859",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TPM Normalisation\n",
    "import pandas as pd\n",
    "\n",
    "# Loading the merged CSV file into a DataFrame\n",
    "df = pd.read_csv('merged_output_file.csv')\n",
    "\n",
    "\n",
    "# Calculating reads per kilobase (RPK)\n",
    "df['RPK'] = df['Mean_Counts'] / (df['Length'] / 1000)\n",
    "\n",
    "# Calculating per million scaling factor\n",
    "scaling_factor = df['RPK'].sum() / 1e6\n",
    "\n",
    "# Calculate TPM\n",
    "df['TPM'] = df['RPK'] / scaling_factor\n",
    "\n",
    "\n",
    "# Defining the path for the DataFrame with TPM values\n",
    "output_tpm_path = 'Normalized_shScramble_output_file_with_TPM4.csv'\n",
    "\n",
    "# To Save the DataFrame to a new CSV file\n",
    "df.to_csv(output_tpm_path, index=False)\n",
    "\n",
    "print(f\"Data with TPM normalization saved to {output_tpm_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a2a23dd6-5324-4f32-9961-67a93268d2e8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('Normalized_shScramble_output_file_with_TPM4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b80315a1-1084-49bf-a926-7fcb71ecbc52",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Length</th>\n",
       "      <th>Replicate1</th>\n",
       "      <th>Replicate2</th>\n",
       "      <th>Mean_Counts</th>\n",
       "      <th>RPK</th>\n",
       "      <th>TPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>2618</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>4172</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.239693</td>\n",
       "      <td>1.146767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>5540</td>\n",
       "      <td>145</td>\n",
       "      <td>139</td>\n",
       "      <td>142</td>\n",
       "      <td>25.631769</td>\n",
       "      <td>122.630332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20044</th>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>297</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20045</th>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>1378</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20046</th>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>1812</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20047</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>525</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20048</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>1141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20049 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Geneid  Length  Replicate1  Replicate2  Mean_Counts  \\\n",
       "0       ENSG00000186092.7    2618           0           0            0   \n",
       "1       ENSG00000284733.2     939           0           0            0   \n",
       "2       ENSG00000284662.2     939           0           0            0   \n",
       "3      ENSG00000187634.13    4172           2           1            1   \n",
       "4      ENSG00000188976.11    5540         145         139          142   \n",
       "...                   ...     ...         ...         ...          ...   \n",
       "20044   ENSG00000212907.2     297           0           0            0   \n",
       "20045   ENSG00000198886.2    1378           0           0            0   \n",
       "20046   ENSG00000198786.2    1812           0           0            0   \n",
       "20047   ENSG00000198695.2     525           0           0            0   \n",
       "20048   ENSG00000198727.2    1141           0           0            0   \n",
       "\n",
       "             RPK         TPM  \n",
       "0       0.000000    0.000000  \n",
       "1       0.000000    0.000000  \n",
       "2       0.000000    0.000000  \n",
       "3       0.239693    1.146767  \n",
       "4      25.631769  122.630332  \n",
       "...          ...         ...  \n",
       "20044   0.000000    0.000000  \n",
       "20045   0.000000    0.000000  \n",
       "20046   0.000000    0.000000  \n",
       "20047   0.000000    0.000000  \n",
       "20048   0.000000    0.000000  \n",
       "\n",
       "[20049 rows x 7 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0446354-0c30-4315-beeb-854555da1860",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Upon processing the data for all four conditions (shScramble, shDDX41, CX5461, and DMSO), we constructed the final dataset. \n",
    "#The dataset structure includes the first column populated with 20,049 gene IDs. Columns two through 140 contain RBP signal data corresponding to each gene ID. \n",
    "#The last four columns represent the FeatureCounts of these genes, which have been normalized using TPM,\n",
    "#ensuring accurate representation of gene expression levels across different samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11a84b0f-f066-43bc-ae36-fc2980a62773",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First step to add RBPs data merged\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Load the main file containing Geneid\n",
    "featurecounts_path = 'chromosomes_only/Featurecounts_proteinfiltered_length_output_file.csv'\n",
    "featurecounts_df = pd.read_csv(featurecounts_path)\n",
    "\n",
    "# Initialize a dictionary to hold data for the final DataFrame\n",
    "final_data = {'Geneid': featurecounts_df['Geneid'].values}\n",
    "columns_dtype = {}\n",
    "\n",
    "rbp_folder_path = 'chr-summed_output_rbp'\n",
    "rbp_files = [f for f in os.listdir(rbp_folder_path) if f.endswith('.csv')]\n",
    "\n",
    "# Pre-fill the dictionary with Geneids as keys and a list full of zeros as values\n",
    "for rbp_file in rbp_files:\n",
    "    column_name = rbp_file.replace('.csv', '')\n",
    "    final_data[column_name] = np.zeros(len(featurecounts_df), dtype=float)\n",
    "    columns_dtype[column_name] = 'float64'\n",
    "\n",
    "# Process each RBP file\n",
    "for rbp_file in rbp_files:\n",
    "    rbp_path = os.path.join(rbp_folder_path, rbp_file)\n",
    "    rbp_df = pd.read_csv(rbp_path)\n",
    "    column_name = rbp_file.replace('.csv', '')\n",
    "    \n",
    "    # Map 'signalValue's to 'Geneid'\n",
    "    for i, gene_id in enumerate(final_data['Geneid']):\n",
    "        value = rbp_df.loc[rbp_df['gene_id'] == gene_id, 'signalValue'].sum()  # Sum in case of multiple entries\n",
    "        if not pd.isnull(value):\n",
    "            final_data[column_name][i] = value\n",
    "\n",
    "# Create the final DataFrame\n",
    "final_df = pd.DataFrame(final_data)\n",
    "\n",
    "# Specify dtype explicitly to avoid FutureWarning\n",
    "for col, dtype in columns_dtype.items():\n",
    "    final_df[col] = final_df[col].astype(dtype)\n",
    "\n",
    "# Save the final DataFrame to a new CSV file\n",
    "final_output_path = 'best_finallllllll_output2.csv'\n",
    "final_df.to_csv(final_output_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31373a4f-3e1e-4982-8f14-dc95d3c56950",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Then to add last 4 columns (translation outcome for each condition)\n",
    "import pandas as pd\n",
    "\n",
    "# Load the final output file\n",
    "final_df = pd.read_csv('best_finallllllll_output_April.csv')\n",
    "\n",
    "conditions = ['shScramble', 'shDDX41', 'DMSO', 'CX5461']\n",
    "condition_files = [\n",
    "    'chromosomes_only/shScramble/Normalized_shScramble_output_file_with_TPM4.csv',\n",
    "    'chromosomes_only/shDDX41/Normalized4_shDDX41_output_file_with_TPM.csv',\n",
    "    'chromosomes_only/DMSO/Normalized4_DMSO_output_file_with_TPM.csv',\n",
    "    'chromosomes_only/CX5461/Normalized_CX5461_output_file_with_TPM.csv',\n",
    "]\n",
    "\n",
    "for condition, file_path in zip(conditions, condition_files):\n",
    "    # Load the condition file\n",
    "    condition_df = pd.read_csv(file_path)\n",
    "    \n",
    "    # Extract Geneid and TPM columns (assuming the column is named 'TPM'; adjust if necessary)\n",
    "    tpm_data = condition_df[['Geneid', 'TPM']].rename(columns={'TPM': f'TPM_{condition}'})\n",
    "    \n",
    "    # Merge the TPM data into the final DataFrame\n",
    "    final_df = pd.merge(final_df, tpm_data, on='Geneid', how='left')\n",
    "\n",
    "# Replace NaN values with 0 (assuming you want to treat missing TPM data as 0)\n",
    "final_df.fillna(0, inplace=True)\n",
    "\n",
    "\n",
    "final_df.to_csv('ready_file_for_ML_April.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "bcf217a4-7a21-4545-8509-14339a9ea843",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "visual = pd.read_csv('ready_file_for_ML_April.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "1b8facf0-9958-4734-ad7b-c50401ee05b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>DDX1</th>\n",
       "      <th>XRCC6</th>\n",
       "      <th>GEMIN5</th>\n",
       "      <th>DROSHA</th>\n",
       "      <th>HNRNPUL1</th>\n",
       "      <th>FTO</th>\n",
       "      <th>MORC2</th>\n",
       "      <th>SSB</th>\n",
       "      <th>U2AF2</th>\n",
       "      <th>...</th>\n",
       "      <th>PPIL4</th>\n",
       "      <th>EEF2</th>\n",
       "      <th>SUPV3L1</th>\n",
       "      <th>DDX6</th>\n",
       "      <th>UPF1</th>\n",
       "      <th>TIA1</th>\n",
       "      <th>shScramble</th>\n",
       "      <th>shDDX41</th>\n",
       "      <th>DMSO</th>\n",
       "      <th>CX5461</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.146767</td>\n",
       "      <td>1.151633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.662869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.674923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>122.630332</td>\n",
       "      <td>137.026864</td>\n",
       "      <td>179.017168</td>\n",
       "      <td>164.731406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20044</th>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20045</th>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20046</th>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20047</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.606961</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.549403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20048</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.665060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20049 rows Ã— 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Geneid  DDX1  XRCC6  GEMIN5  DROSHA  HNRNPUL1  FTO  MORC2  \\\n",
       "0       ENSG00000186092.7   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "1       ENSG00000284733.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "2       ENSG00000284662.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "3      ENSG00000187634.13   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "4      ENSG00000188976.11   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "...                   ...   ...    ...     ...     ...       ...  ...    ...   \n",
       "20044   ENSG00000212907.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "20045   ENSG00000198886.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "20046   ENSG00000198786.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "20047   ENSG00000198695.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "20048   ENSG00000198727.2   0.0    0.0     0.0     0.0       0.0  0.0    0.0   \n",
       "\n",
       "       SSB  U2AF2  ...     PPIL4  EEF2    SUPV3L1  DDX6      UPF1  TIA1  \\\n",
       "0      0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "1      0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "2      0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "3      0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "4      0.0    0.0  ...  0.000000   0.0   0.000000   0.0  3.674923   0.0   \n",
       "...    ...    ...  ...       ...   ...        ...   ...       ...   ...   \n",
       "20044  0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "20045  0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "20046  0.0    0.0  ...  0.000000   0.0   0.000000   0.0  0.000000   0.0   \n",
       "20047  0.0    0.0  ...  3.606961   0.0  44.549403   0.0  0.000000   0.0   \n",
       "20048  0.0    0.0  ...  0.000000   0.0   4.665060   0.0  0.000000   0.0   \n",
       "\n",
       "       shScramble     shDDX41        DMSO      CX5461  \n",
       "0        0.000000    0.000000    0.000000    0.000000  \n",
       "1        0.000000    0.000000    0.000000    0.000000  \n",
       "2        0.000000    0.000000    0.000000    0.000000  \n",
       "3        1.146767    1.151633    0.000000    0.662869  \n",
       "4      122.630332  137.026864  179.017168  164.731406  \n",
       "...           ...         ...         ...         ...  \n",
       "20044    0.000000    0.000000    0.000000    0.000000  \n",
       "20045    0.000000    0.000000    0.000000    0.000000  \n",
       "20046    0.000000    0.000000    0.000000    0.000000  \n",
       "20047    0.000000    0.000000    0.000000    0.000000  \n",
       "20048    0.000000    0.000000    0.000000    0.000000  \n",
       "\n",
       "[20049 rows x 144 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c3e14f4-d706-46e4-9f64-14d7b299a31c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "tf2-cpu.2-11.m117",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/tf2-cpu.2-11:m117"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
